### ISSUE TITLE
Question on how to apply on Ray-marching

### ISSUE THREAD

**Issue by Zhentao-Liu** [Ezio77/DINER#7]
Can I use ray marching to train my nerf model based on DINER? After I sample points, I just pass them into the _interp model variants to get the results? BTW, Could you give a more detailed explaination about (d) hash-mapped input (f) Learned INR in DINER in Fig 2? I do not understand how to get them. 
Best regards!

**Comment by Ezio77** [comment]
Thanks for your interest in our work. Sure you can use the _interp model to obtain values by sending the sampling points and then perform ray marching. 
        Regarding the hash mapped input, each point (x', y') in the image corresponds to a value obtained after applying hash coding to the original image coordinates (x, y). The color value of each point is predicted from the diner model.
        In order to obtain the hashed input (d), after training DINER, you can iterate through each value in the hash table, feed it into the MLP to obtain the corresponding predicted color value, and then plot it. 
        As for the learned INR, we mainly focus on the properties of the second half of the diner model(MLP). We feed normalized grid coordinates into the trained MLP and visualize the learned function mapping, which represents a mapping relationship from the hashed coordinates to their corresponding colors.
        I hope my answer can help you.

**Comment by Zhentao-Liu** [comment]
> ```
>     Thanks for your interest in our work. Sure you can use the _interp model to obtain values by sending the sampling points and then perform ray marching. 
>     Regarding the hash mapped input, each point (x', y') in the image corresponds to a value obtained after applying hash coding to the original image coordinates (x, y). The color value of each point is predicted from the diner model.
>     In order to obtain the hashed input (d), after training DINER, you can iterate through each value in the hash table, feed it into the MLP to obtain the corresponding predicted color value, and then plot it. 
>     As for the learned INR, we mainly focus on the properties of the second half of the diner model(MLP). We feed normalized grid coordinates into the trained MLP and visualize the learned function mapping, which represents a mapping relationship from the hashed coordinates to their corresponding colors.
>     I hope my answer can help you.
> ```

many thanks, I do in this way. But I get everywhere zeros :<, the input corrds should be normalized into range of [-1, 1] right?

**Comment by Ezio77** [comment]
Yes, if you want to obtain the learned INR, you need to evaluate the range of values in the hash table after training DINER. For images, the range is typically around [-0.3, 0.3]. You need to pay attention to the maximum and minimum values and normalize them to [-1, 1].

**Comment by Ezio77** [comment]
You can use render_hash_image() in utils.py to generate a learned INR after you trained DINER to fit an image.

**Comment by Zhentao-Liu** [comment]
Thx for your adivce, I have got hash input and learned INR images. 
When I apply DINER to my task, it is quite wired. 
Could you leave you email here so I can connect you in private.

**Comment by Ezio77** [comment]
Sure, my email is mg21230085@smail.nju.edu.cn.
